scrapy_dropbox/__init__.py

```diff
@@ -1,46 +1,98 @@
-import dropbox
-import typing
+import logging
 import os
 
+from dropbox import Dropbox, files
+from dropbox.exceptions import ApiError
+from dropbox.files import WriteMode
 from scrapy.exceptions import NotConfigured
 from scrapy.extensions.feedexport import BlockingFeedStorage
-from dropbox.files import WriteMode
+
+from .utils import parse_uri
+
+logger = logging.getLogger(__name__)
+CHUNK_SIZE = 4 * 1024 * 1024
 
 
 class DropboxFeedStorage(BlockingFeedStorage):
+    def __init__(self, uri, api_token, *, feed_options):
```

Comment:
This is a backward-incompatible change, i.e. if someone were using a `DropboxFeedStorage` subclass calling `super().__init__(uri)`, their code would break.

You can make it backward-compatible by making `from_crawler` have a fallback if the `__init__` method does not work with all 3 parameters, and pass only the uri then, and log a deprecation warning. There are a few examples in the Scrapy code base I think.

Reply:
Thanks @Gallaecio , great point!
**question:** In the `_store_in_thread()` I had added some modifications. Hence some of the code there is dependent on the changes I did in the `__init__()`. (e.g: credential validation and uri validation). Shall i revert those changes then? Because for maintaining backward-compatibility, I think that would also be necessary. Let me know what you think about it.

Reply:
The idea is that, even if the subclass does not support the new parameter and is called without it, you then edit the resulting object in `from_crawler` to make up for it, so that when `_store_in_thread()` is called it always has all the expected attributes set as expected. Something like this:

```python
def _init_from_crawler(storage, crawler, feed_options):
    ...
    storage.dropbox_path = dropbox_path
    ...


class DropboxFeedStorage(BlockingFeedStorage):

    @classmethod
    def from_crawler(cls, crawler, uri, *, feed_options=None):
        try:
            return cls(uri, crawler, feed_options)
        except TypeError:
            warn(
                (
                    "Defining subclasses of DropboxFeedStorage that do not "
                    "accept the crawler and feed_options parameters in their "
                    "__init__ method is deprecated, and will stop working in "
                    "a future version."
                ), 
                DeprecationWarning,
            )
            storage = cls(uri)
            _init_from_crawler(storage, crawler, feed_options)
            return storage        
        
    def __init__(self, uri, crawler=None, feed_options=None):
        self.uri = uri
        if crawler:
            _init_from_crawler(self, crawler, feed_options)
```

I know it is messy, but it should be a temporary mess. After some deprecation period (e.g. 1 year as we do in Scrapy), you can switch to the clean version.

Reply:
@Gallaecio , Apologies for the wait on this and thank you so much for the advice. That indeed makes sense and I've made those changes.